File: 362__Design_Hit_Counter.py
class HitCounter(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.data = []
        self.start = 0


    def hit(self, timestamp):
        """
        Record a hit.
        @param timestamp - The current timestamp (in seconds granularity).
        :type timestamp: int
        :rtype: void
        """
        self.data.append(timestamp)


    def getHits(self, timestamp):
        """
        Return the number of hits in the past 5 minutes.
        @param timestamp - The current timestamp (in seconds granularity).
        :type timestamp: int
        :rtype: int
        """
        while self.start < len(self.data) and timestamp - self.data[self.start] >= 300:
            self.start += 1
        if self.start == len(self.data):
            return 0
        return len(self.data) - self.start




# Your HitCounter object will be instantiated and called as such:
# obj = HitCounter()
# obj.hit(timestamp)
# param_2 = obj.getHits(timestamp)

File: Java/362__Design_Hit_Counter.java
class HitCounter {

    Deque<Integer> record;

    /** Initialize your data structure here. */
    public HitCounter() {
        record = new LinkedList();
    }

    /** Record a hit.
        @param timestamp - The current timestamp (in seconds granularity). */
    public void hit(int timestamp) {
        checkRecord(timestamp);
        record.addLast(timestamp);
    }

    /** Return the number of hits in the past 5 minutes.
        @param timestamp - The current timestamp (in seconds granularity). */
    public int getHits(int timestamp) {
        checkRecord(timestamp);
        return record.size();
    }

    private void checkRecord(int timestamp){
        if (timestamp > 300) {
            int start = timestamp - 300 + 1;
            while (record.size() > 0 && record.peekFirst() < start)
                record.poll();
        }
    }
}

/**
 * Your HitCounter object will be instantiated and called as such:
 * HitCounter obj = new HitCounter();
 * obj.hit(timestamp);
 * int param_2 = obj.getHits(timestamp);
 */
